# NumPy Basics

Short for Numerical Python, it is one of the most important foundational packages for numerical computing in Python. You will very likely encounter this package throughout your software experience, including AI methods such as machine learning. NumPy provides:

-np.array: a function to convert objects of Python into "ndarrays", which are the Python equivalent of a Matlab matrix

-array functions and methods that make for easy manipulation of matrices as seen in linear algebra, along with operations one takes for granted on matrices, such as addition, multiplication, etc. 

-NumPy provides functions that can easily operate on matrices, for vectorizing code or applying function on a by-element basis, as Matlab does (instead of applying LOOPS to each element of a matrix). For instance, the Python math module called as **import math** has its own natural logarithm function, math.log(). But this logarithm function does not overload to arrays. NumPy has np.log() which allows you to operate on arrays, where if A is an array, np.log(A) applies the natural log to each element of A. You will see more functions as we proceed. 

-other linear algebra methods and algorithms

-other capabilities, such as random number generation, Fourier transforms, etc

 # SciPy

Let us mention quickly SciPy, which is a collection of packages for some standard applications used in scientific computing. It includes
* scipy.integrate: Numerical integration routines and differential equation solvers
* scipy.linalg: Linear algebra routines and matrix decompositions extending beyond what NumPy offers in 
numpy.linalg
* scipy.optimize: function optimizers and root finding algorithms
* scipy.stats: standard continuous and discrete probability distribution functions, statistical tests, etc

 


# Import numpy as np

We write **import numpy as np** to call numpy functions using np.*Numpy_function_name_here*(). You could write **import numpy as WHATEVER** and call, say the log function as WHATEVER.log(), but it has become custom to use "np". There are ways to import numpy functions/methods without using any pre-qualifier like 'np.log' and just use 'log', but this is not recommended, since many numpy functions have the same name as other Python functions or those in other modules you might import, creating potential naming conflicts. An example is the logarithm function, which math.log() calls the log from the math module, and np.log() calls the log function from the NumPy module. 

In [1]:
import numpy as np


# Python lists vs np.array: An introduction to Python containers

 Python has by default a list structure, which can be assigned, an its respective elements accessed using the following example syntax (note Python is zero-based in its indices, unlike Matlab) 

In [2]:
a_list=[2,0,1]
print(a_list)
print(a_list[2])
print('=========================')

for i in range(3):
    print(a_list[i])

[2, 0, 1]
1
2
0
1


**Python lists are limited!** Note Python lists are not suitable objects for operating on them like vectors are, you can reassign values to lists, but you cannot add 2 lists:

In [3]:
a_list[0]=-2
print(a_list)
another_list=[-1, -10, 0]

print(a_list+another_list)  #<---does not work like vector addition does! (Instead concatenates lists)

[-2, 0, 1]
[-2, 0, 1, -1, -10, 0]


Instead, thanks to the "import numpy as np" command we typed into one of the above cells, we can **from a list**, obtain a "true vector", called a "numpy array" or "ndarray" (We will use the words 'numpy arrays', 'NumPy arrays', 'ndarrays' interchangebly). We use the np.array() function on a list to convert it into an array, and now they act like vectors:

In [4]:
vector_1=np.array(a_list)
vector_2=np.array(another_list)
print(vector_1+2*vector_2)

[ -4 -20   1]


**What about matrices?** Similar to "raw Python lists" vs numpy array structures, we can note that the following matrix-looking object does not really behave like a matrix:

In [5]:
array_looking_thing=[[1,2,3],
                     [4,5,6],
                     [1,0,0]]  #list of lists

print(2*array_looking_thing)

[[1, 2, 3], [4, 5, 6], [1, 0, 0], [1, 2, 3], [4, 5, 6], [1, 0, 0]]


In [6]:
array=np.array(array_looking_thing) #numpy array/ ndarray
print(2*array)

[[ 2  4  6]
 [ 8 10 12]
 [ 2  0  0]]


In [7]:
list2=[2, 'a', 23]
print(list2[1])

a


The Numpy array has many more capabilities than the above shown. Also, these NumPy arrays, also known as *ndarrays*, use a more efficient internal memory representation and save memory use. 

It is important to know that in programming languages, under the hood, manipulating arrays is not such a trivial memory operation as it would seem. (Try finding what similar structure/object, to say, 2x2 arrays (of Python, Matlab, etc) are used in C++!)

On the other hand, you can overload NumPy functions to numpy arrays. Let us do the following, we import the math module and use its natural log function, which works well on scalar numbers, **but not on higher dimension ndarrays:**

In [8]:
import math  # <---includes standard math functions such as log, trig functions
math.log(3)

1.0986122886681098

In [9]:
list=[3,9,1]
array=np.array(list)
math.log(array)
math.log(list)


TypeError: only size-1 arrays can be converted to Python scalars

But the LOGARITHM function within the numpy library, namely np.log(), works perfectly well with both naked lists and numpy arrays:

In [10]:
 print(np.log(array))  # <----numpy version of the log function, called as np.log (np.log vs math.log)
print(np.log(list))  # <---np.log() can operate on arrays/lists,etc


[1.09861229 2.19722458 0.        ]
[1.09861229 2.19722458 0.        ]


In [11]:
# As seen above, we can create arrays from lists and then using np.array() on the 
# list to convert it into a numpy array. Once again let us do this:
list=[[1,2,3], [4,5,6], [1,0,0]]
arr=np.array(list)  #initialize an array from the list
print(arr)

[[1 2 3]
 [4 5 6]
 [1 0 0]]


We can access the dimensions and their respective values, of the numpy arrays, as follows:

In [12]:

print('We have that arr.dim=', arr.ndim)   
print(arr.shape)  #this gives the actual values of these two dimensions

We have that arr.dim= 2
(3, 3)


Either of np.identity( *size* ) or np.eye( *size* ) create a *size* $\times$ *size* identity matrix:

In [13]:
A=np.eye(4)
B=np.identity(3)
print(A)
print('\n')
print(B)

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


 Other ways to create numpy arrays, and populate them through, say a FOR-LOOP are through the np.zeros((rows,col)) commands or np.zeros(N). Another way is np.ones()

In [14]:
A=np.zeros((3,1))
A[1,0]=1
print(A)
B=np.zeros((3,5))
print('\n', B, '\n') #<---Note separate by a comma two different things to print!

N=17
C=np.zeros(N)
for k in range(N):
    C[k]=np.log(k+1)

print(C)    

D=np.ones((3,3))
print('\n',D)

[[0.]
 [1.]
 [0.]]

 [[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]] 

[0.         0.69314718 1.09861229 1.38629436 1.60943791 1.79175947
 1.94591015 2.07944154 2.19722458 2.30258509 2.39789527 2.48490665
 2.56494936 2.63905733 2.7080502  2.77258872 2.83321334]

 [[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [15]:
C.shape

(17,)

In [16]:
C.ndim

1

In [17]:
C.T

array([0.        , 0.69314718, 1.09861229, 1.38629436, 1.60943791,
       1.79175947, 1.94591015, 2.07944154, 2.19722458, 2.30258509,
       2.39789527, 2.48490665, 2.56494936, 2.63905733, 2.7080502 ,
       2.77258872, 2.83321334])

In [18]:
C

array([0.        , 0.69314718, 1.09861229, 1.38629436, 1.60943791,
       1.79175947, 1.94591015, 2.07944154, 2.19722458, 2.30258509,
       2.39789527, 2.48490665, 2.56494936, 2.63905733, 2.7080502 ,
       2.77258872, 2.83321334])

In [19]:
C2=C.reshape(-1,1)
print(C2)


[[0.        ]
 [0.69314718]
 [1.09861229]
 [1.38629436]
 [1.60943791]
 [1.79175947]
 [1.94591015]
 [2.07944154]
 [2.19722458]
 [2.30258509]
 [2.39789527]
 [2.48490665]
 [2.56494936]
 [2.63905733]
 [2.7080502 ]
 [2.77258872]
 [2.83321334]]


In [ ]:
print(C.ndim)
print(C2.T)

In [34]:
W=C[:,np.newaxis]
print('W=', W, '\n')
print('W.shape=', W.shape)
W2=C[np.newaxis,:]
print('STEP1: W2=', W2, '\n')
print('W2.shape=', W2.shape, '\n')
W2=W2.reshape(17,1)
print('STEP2: Now W2=', W2, ' with W2.shape=', W2.shape)
 

W= [[0.        ]
 [0.69314718]
 [1.09861229]
 [1.38629436]
 [1.60943791]
 [1.79175947]
 [1.94591015]
 [2.07944154]
 [2.19722458]
 [2.30258509]
 [2.39789527]
 [2.48490665]
 [2.56494936]
 [2.63905733]
 [2.7080502 ]
 [2.77258872]
 [2.83321334]] 

W.shape= (17, 1)
STEP1: W2= [[0.         0.69314718 1.09861229 1.38629436 1.60943791 1.79175947
  1.94591015 2.07944154 2.19722458 2.30258509 2.39789527 2.48490665
  2.56494936 2.63905733 2.7080502  2.77258872 2.83321334]] 

W2.shape= (1, 17)
STEP2: Now W2= [[0.        ]
 [0.69314718]
 [1.09861229]
 [1.38629436]
 [1.60943791]
 [1.79175947]
 [1.94591015]
 [2.07944154]
 [2.19722458]
 [2.30258509]
 [2.39789527]
 [2.48490665]
 [2.56494936]
 [2.63905733]
 [2.7080502 ]
 [2.77258872]
 [2.83321334]]  with W2.shape= (17, 1)


In [35]:
W2=W2.reshape(1,-1)
print('STEP 3: Now W2.shape=', W2.shape)
W2

STEP 3: Now W2.shape= (1, 17)


array([[0.        , 0.69314718, 1.09861229, 1.38629436, 1.60943791,
        1.79175947, 1.94591015, 2.07944154, 2.19722458, 2.30258509,
        2.39789527, 2.48490665, 2.56494936, 2.63905733, 2.7080502 ,
        2.77258872, 2.83321334]])

# For loop
The general form of a Python for statement/loop is:

**for** *variable* **in** *sequence* **:**

          CODE BLOCK HERE
          
 Where to give a quick example:

           

In [ ]:
print(C2.T)

In [ ]:
 total=0
 for num in (77,11,3):
     total=total+num
 print(total)

# Note that indentation matters in Python! 

This does not happen say in C++ or Matlab, but Python can be a little special/picky about indentation. Note how the indentation in the above for loop was carried out. Try running instead:

In [ ]:
total=0
for num in (77,11,3):
     total=total+num
     print(total)

and note how putting the "print" statement within a slightly deeper indentation, makes it print the running sum at each iterate of the FOR loop. This is supposed to be a notational convenience to make Python easily portable and shareable when working with many developers and users on the same code.  

The expression (77,11,3) is called a **tuple** in Python, and it is different from a **Python list**, /[77, 11, 3/] where tuples are in between parentheses (*tuple elements here*), and lists are in square brackets \[ *list contents here* \]. At the moment this difference is of little relevance to us, but it is useful to keep it in mind, since lists in Python are *mutable*, whereas tuples are *immutable*. 

The above code snippet would run apparently the same though, if we had used a list instead of a tuple:

In [ ]:
total=0
for num in [77,11,3]:
     total=total+num
print(total)

# Using range()

Python has the range function which creates an object with values over which we can iterate:

In [ ]:

for i in range(10): # REMEMBER PYTHON IS ZERO-based!
    print('I like candy', i)

Note that range() is a strange function, because it does 
range(*begin, end, jump size*) **(with default begin=0, and default jump size=1)** but it does not reach the end, as you could see in the previous FOR loop, it did not get to *end*=10, but rather, to *end*-1=9. That's why:

In [ ]:
range(0,10,2)==range(0,9,2)  # <--Evaluates to TRUE! 0,2,4,6,8

In [ ]:
x=range(10)
type(x)

In [ ]:
x=np.linspace(0,10, 11)
print(x)
print(type(x))

We can have negative jump sizes, so that we can have stuff like the following:

In [ ]:
for i in range(10,1, -1):
    print('i=', i)

And once again, note how END is not reached with a -1=*jump size*. This is quite confusing about Python, so get used to it. range() DOES respect ORDER though, so even if the same numbers indices are involved (namely, 0, 2, 4, 6 vs 6, 4, 2, 0) the following evaluates to FALSE:

In [ ]:
range(0,7,2)==range(6,-1, -2)

# Array indexing
We now show some of the basic ways to access elements of NumPy arrays. Take special note of where indices begin and end when they are selected.

In [ ]:
# Illustrate np.arange(), which is like range() but for array creation
arr=np.arange(10)
print(arr)

In [ ]:
arr[0]

In [ ]:
arr[6:9]  # <---note does not include upper index, element arr[9] when displaying below:

In [ ]:
arr[0:2]=13  #re-assignment of a subset of the array's elements 
print(arr)

In [ ]:
A=np.eye(7)
A[6,0]=23   
print(A)

## Array Slicing
Using notation like arr\[6:9\] generates a subslice of the array arr from its indices \[6,9) or 6,7,8. For a two-dimensional array, we show what slicing can do:

In [ ]:
arr2d=np.array([[1,2,3],[4,5,6],[7,8,9]] )
print(arr2d)

Note that these are equivalent ways to access this array's elements:

In [ ]:
arr2d[0][1]

In [ ]:
arr2d[0,1]

This is because

In [ ]:
arr2d[0]

So that arr2d\[0\]\[1\] accesses element 1 of arr2d\[0\]. Now a bit more slicing techniques:

In [ ]:
print('arr2d[1, :]=', arr2d[1, :])
print('arr2d[1,:].ndim=', arr2d[1, :].ndim, '\n')
y=arr2d[1, :].reshape(-1,1)
print(y)
x=arr2d[:,1]

In [ ]:
np.matmul(arr2d[0], arr2d[1])

The previous means "all rows of column 1", which **it prints as a row vector**. You can transpose an array using arr2d.T. Print it out to see. In the case of a vector, it displays it as a row vector even after transposing though. 

In [ ]:
print(arr2d.T, '\n')
vector=arr2d[:, 1]
print(vector.T, '\n')

#To actually reshape into a column vector: 
z=vector.reshape(-1,1) 
vector.shape
z.shape
print(z)

Note the use of indexing like:

In [ ]:
arr2d[:2, 1:]

# Norms in Python
Either of SciPy or NumPy contain methods to compute vector and matrix norms

In [ ]:
import numpy.linalg as LA # <---load numpy functions for linear algebra 

a_vector=[1,-2,4,0,0,2]
a_vector=np.array(a_vector)
print(a_vector)

In [ ]:
print('LA.norm(a_vector)=', LA.norm(a_vector)) #numpy.linalg.norm() function with default Euclidean norm
print('LA.norm(a_vector,2)=', LA.norm(a_vector,2))
print('LA.norm(a_vector,1)=', LA.norm(a_vector,1))
print('LA.norm(a_vector,np.inf)=', LA.norm(a_vector, np.inf))

# About numerical accuracy on computers

Care must be taken when one starts to learn how to implement numerical analysis techniques on a computer and program algorithms, because some basic mathematical rules we take for granted, are not so straightforward on a computer. It all comes down to the way in which numbers are represented in a computer, a finite digit machine. 

In [ ]:
x=0.0
for i in range(10):
    x=x+0.1
    if x==1.0:  #careful when running "equality tests"
        print(x, '=1.0')
    else:
        print(x, 'is not 1.0')
        print(x==10.0*0.1)
            

The previous code snippet is saying that, to the Python interpreter, it is false that 
$\underbrace{0.1+0.1+\cdots +0.1}_{10\;\text{times}}=10.0*(0.1)$! For the record, note that to the Python shell, running 10.0*(0.1) gives

In [ ]:
number=10.0*(0.1)
print(number)

This simple example shows how we cannot take for granted that what we learned in elementary school math holds on a computer. For the Python interpreter, adding ten times 0.1, is not the same as multiplying 0.1 by 10. 

Of course, before you decide to quit numerical analysis, or lose hope in programming with numbers involved, we must try to understand what is happening here.

The lesson here is that **one must worry about tests for numerical EQUALITY**. As seen above, using == to compare two floating point numbers can produce unintuitive results, and it can be easy for the non-experienced to forget these simple nuances.

It is almost always more appropriate to ask whether two floating point numbers are **close enough to each other, and not if they are "identical"**. It's better to ask if abs(x-y)<0.0001, rather than if x==y. 

In [ ]:
x=0  #just noticing it still happens even if I write x=0 (integer?)
for i in range(10):
    x=x+0.1  #<---I think this line creates the conversion of x to float
    if x==1:
        print(x, '=1.0')
    else:
        print(x, 'is not 1.0')
        print(x==10.0*0.1)

The reason this happens is because of the binary representation of the numbers in the computer. Remember that an integer $x$ can be represented in binary as
$$x=\sum_{j=0}d_j 2^j.$$ For instance, the number 6 is 
$$6=1*2^2+1*2^1+ 0*2^0, $$ which is written in binary as 110, where here 0 is the rightmost bit, corresponding to $0*2^0$, and the leftmost bit of 1 indicates the highest power of 2 involved was $2^2$. 

Now, in normal (to us humans, since elementary school) decimal point notation, we can represent the number 1.949 as the pair $(1949, -3)$ which is shorthand for
$$1.949=1949*10^{-3}$$

The number of significant digits determines the precision with which numbers can be represented. If there were only *two significant digits*, the number 1.949 would not be possible to represent exactly. Best we would be able to do is to approximate it as 1.9, or in the pair notation: $(19, -1)$. This is called the rounded value.

Back to binary notation, the significant digits and exponents (exponent to which we raise 2) are expressed in binary notation. For example, the number 0.625=5/8 is represented as
$$.625\Leftrightarrow (101, -11)$$ because 101 is the binary representation of 5 and -11 the binary representation of -3, so the pair
$$(101, -11)\Leftrightarrow  5*2^{-3}=5/8=0.625$$

# Mini exercise:
* Convince yourself that there do not exist integers $s, e\in\mathbb{Z}$ such that
$$0.1=s\cdot 2^e$$

So no matter how many bits Python, or any other programming language use to represent floating-point numbers, it can only approximate 0.1. If one checks the best approximation Python can do to 0.1, most likely it is slightly larger than 0.1. 

But still, the loop above gave us a number SMALLER than 1.0! This is because Python had to do some additional rounding downwards during the loop.

# Don't lose sleep over the previous

The previous is no reason to become a paranoid though. It only means we have to be a little bit careful. The commonly/widespread used functions or the ones coming in common Python distributions and packages like numpy, scipy, etc have been built to be safe and deal with these numerical problems one can potentially face. Checking a statement like the residue of dividing, say 132 by 11 and 12 and using 132%12==0 and 132%11==0, is a safe thing to do, for instance: 